### ISSUE TITLE
Unable to handle/complete transactions on startup

### ISSUE THREAD

**Issue by robhellstrom** [bizz84/SwiftyStoreKit#197]
### Platform

- [X] iOS
- [ ] macOS
- [ ] tvOS

### In app purchase type

- [ ] Consumable
- [X] Non-consumable 
- [ ] Auto-Renewable Subscription
- [ ] Non-Renewing Subscription

### Environment

- [X] Sandbox
- [ ] Production

### Version
0.7.1

### Related issues
https://github.com/bizz84/SwiftyStoreKit/issues/146

### Issue summary
Using simulator, a number of transactions are queued and the application shut down. On startup, these outstanding transactions remain held in a queue and unable to be processed. I need to be able to understand what are these and handle them/complete them.

### What did you expect to happen
I have a simple non-consumable content IAP. All fair-weather tests have been successful. I am completing bad-weather tests- I am trying to reproduce a boundary case due to a user force-quit immediately after agreeing to purchase an IAP. StoreKit continues to prompt the user in this case to complete the upgrade and in my boundary case the user agrees to purchase. That works with the code I have using SwiftyStoreKit. 

On startup I expect to be able to somehow know that the transaction has completed, and then unlock features/content accordingly. In my particular case none of this relies on downloaded content, so I believe I should use atomically: true.

So I expect to see any unhandled transactions completed and dealt with by code which executes once on startup.

    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        completeIAPTransactions()
        return true
    }
    func completeIAPTransactions() {
        SwiftyStoreKit.completeTransactions(atomically: true) { products in
            for product in products {
                print("\(product.productId) productId transactionId \(product.transaction.transactionIdentifier) transaction state \(product.transaction.transactionState)")
                self.completeStoreKitTransactions(product: product)
            }
        }
    }
    
    func completeStoreKitTransactions(product: Product) {
        
        if !product.needsFinishTransaction {
            return
        }
        
        if product.transaction.transactionState == .purchased  {
            
            if product.needsFinishTransaction {
                
                SwiftyStoreKit.finishTransaction(product.transaction)
            }
            
            // handle purchased
            print("completeStoreKitTransactions: .purchased ")
            
        } else if product.transaction.transactionState == .restored {
            
            if product.needsFinishTransaction {
                
                SwiftyStoreKit.finishTransaction(product.transaction)
            }
            
            // handle restored
            print("completeStoreKitTransactions: .restored ")
        }
    }

### What happened instead
On start up, SwiftyStoreKit.completeTransactions(atomically: true) is executed but none of the completion code is executed so it seems no transactions are left to complete. Just after, however, all unhandled transactions are listed out by the print statement in the SwiftyStoreKit delegate 

func paymentQueue(_ queue: SKPaymentQueue, updatedTransactions transactions: [SKPaymentTransaction]) {

        if unhandledTransactions.count > 0 {
            print("unhandledTransactions: \(unhandledTransactions)")
        }

}

Any attempts to continue to execute new transactions build the payment queue.

**Comment by bizz84** [comment]
First of all, please update to the latest version of SwiftyStoreKit.

> Just after, however, all unhandled transactions are listed out by the print statement in the SwiftyStoreKit delegate

What is the transaction state for the `unhandledTransactions`? If it's `.purchasing`, the transactions are not yet complete. If it's `.purchased`, `.restored`, `.failed` or `.deferred`, these should be returned by `completeTransactions`.

**Comment by robhellstrom** [comment]
Updated to 0.8.6.

According to your code, transaction state is only `.purchasing` : refer to PaymentQueueController.swift in SwiftyStoreKit...

```
// MARK: SKPaymentTransactionObserver
    func paymentQueue(_ queue: SKPaymentQueue, updatedTransactions transactions: [SKPaymentTransaction]) 
```

```
        var unhandledTransactions = paymentsController.processTransactions(transactions, on: paymentQueue)

        unhandledTransactions = restorePurchasesController.processTransactions(unhandledTransactions, on: paymentQueue)

        unhandledTransactions = completeTransactionsController.processTransactions(unhandledTransactions, on: paymentQueue)

        unhandledTransactions = unhandledTransactions.filter { $0.transactionState != .purchasing }
        if unhandledTransactions.count > 0 {
            let strings = unhandledTransactions.map { $0.debugDescription }.joined(separator: "\n")
            print("unhandledTransactions:\n\(strings)")
        }
```
When this runs I see multiple transactions printed from the strings array.

**Comment by zeromhz** [comment]
I'm having the exact same problem. Running on macOS. Same problem happens in the macOS example. Running 0.9.2. I have been unable to figure out any way to execute code when the transaction completes.

**Comment by robhellstrom** [comment]
I believe I am able to close this now. After not being able to reproduce this except within the simulator and after researching the problem and finding that to be the case for others, I submitted the app and I still can't reproduce this problem on the app store in production. However, on testing the code (accidentally) with the transactions set to .production, this unblocked the transactions and they flowed freely into the simulator. I had so many it took some time, but all were honoured, and processed, by the  SwiftyStoreKit. I am impressed by the SwiftyStoreKit. For someone new to IAP this was a quick solution for a final stage development before going into production. I can recommend it.
